kube-aws-updater

```diff
@@ -0,0 +1,184 @@
+    done
+  fi
+}
+
+label_for_cycling() {
+  local role=$1
+  local nodes=$(kubectl --context=${kube_context} get nodes -l role=${role} -o json | jq -r '.items[].metadata.name')
+  echo "${kube_context} :: $(echo "${nodes}" | wc -l) ${role}s"
+  echo "Labelling ${role}s for retirement..."
+  for node in ${nodes}; do
+    kubectl --context=${kube_context} label node ${node} retiring=${retire_time} --overwrite=true
+    kubectl --context=${kube_context} cordon ${node}
+  done
+}
+
+kill_node() {
+  local node=$1
+
+  echo "---------- ${node} ----------"
+  echo "draining ${node}..."
+  time kubectl --context=${kube_context} drain ${node} --ignore-daemonsets --force --delete-local-data
+  local instance_id=$(aws ${aws_opts} --output=json ec2 describe-instances --filters "Name=network-interface.private-dns-name,Values=${node}" \
+    | jq -r '.Reservations[].Instances[].InstanceId')
+  echo "terminating ${node} with instance-id ${instance_id}"
+  time aws ${aws_opts} ec2 terminate-instances --instance-ids=${instance_id}
+  echo "----------------------------------------------------------------------"
+}
+
+cycle_nodes() {
+  local role=$1
+
+  # - double ASG size
+  local instance_address=$(kubectl --context=${kube_context} get nodes -l role=${role},retiring=${retire_time} -o json | jq -r '.items[0].metadata.name')
+  local instance_id=$(aws ${aws_opts} --output=json ec2 describe-instances --filters "Name=network-interface.private-dns-name,Values=${instance_address}" \
+    | jq -r '.Reservations[].Instances[].InstanceId')
+  local asg_name=$(aws ${aws_opts} --output=json autoscaling describe-auto-scaling-groups \
+    | jq -r '."AutoScalingGroups"[] | select(."Instances"[]."InstanceId"=="'${instance_id}'") | ."AutoScalingGroupName"')
+  local asg_count=$(aws ${aws_opts} --output=json autoscaling describe-auto-scaling-groups \
+    | jq -r '."AutoScalingGroups"[] | select(."Instances"[]."InstanceId"=="'${instance_id}'") | ."DesiredCapacity"')
+  aws ${aws_opts} --output=json autoscaling update-auto-scaling-group --auto-scaling-group-name "${asg_name}" --desired-capacity $(( ${asg_count} * 2)) --max-size $(( ${asg_count} * 2))
+
+  # - wait for 2x group
+  local count=0
+  while [[ ${count} -lt ${asg_count} ]]; do
+    sleep 8
+    count=$(kubectl --context=${kube_context} get node -lrole=${role} -Lrole,retiring \
+      | grep -v ${retire_time} \
+      | grep -v NotReady \
+      | tail -n +2 \
+      | wc -l)
+    echo "${count}/${asg_count} ready ${role}s"
+  done
+
+  # - stop ASG actions
+  aws ${aws_opts} --output=json autoscaling suspend-processes --auto-scaling-group-name "${asg_name}"
+
+  ### - move the registry
+  if [[ "${role}" == "worker" ]]; then
+    handle_registry
+  fi
+
+  # - drain/terminate labelled masters (sequentially)
+  local old_nodes=$(kubectl --context=${kube_context} get nodes -l role=${role},retiring=${retire_time} -o json | jq -r '.items[].metadata.name')
+  for old_node in ${old_nodes}; do
```

Comment:
Quotes?

Reply:
Do I want quotes here, that will mean `${old_nodes}` is not a string, not an array.

Reply:
Resolved